### PR TITLE
Add unit tests for extension

### PR DESCRIPTION
**Pull request by connorff**
## Overview
Adds extension unit tests to `ide/src/tests/`.

### What's tested

1. Forward and backward select highlights correct slice values
2. The extension installs correctly when using the latest Flowistry release
3. The extension installs correctly when using the current Flowistry build

### Implementation details
**Slice tests:**
Slice tests open `mock_project/` in VSCode and execute flowistry commands on values. `slice/mock_data/slices.ts` contains the expected slice values.

**Install tests:**
Install tests load the extension and wait for the VSCode Flowistry commands to be available. Because each test starts from a fresh build without Flowistry binaries, the extension must download the binaries from the latest release on Github. To install from the current Flowistry build, we start a proxy server which replaces requests to Github `.zip`s with a `.zip` of the current build.

### Potential improvements:
- [ ] Add tests for slice text decorations and effects panel
- [ ] Add tests for failing projects (Rust project that doesn't compile)
- [x] Split the test job from `tests.yml` into multiple jobs for building and installing the rust project, extension, and testing the extension

**Note**: tests run fine in Github Actions, but I had to add a ~5 second delay before `vscode.commands.executeCommand("flowistry.forward_select")` in `extension.test.ts` to avoid an error with `flowistry.forward_select` not existing as a vscode command. ~~Still finding the root cause of this, but would appreciate seeing if others experience the same issue locally.~~ **Update:** For some reason downgrading `npm` to v6 fixed this.

### CODE REVIEW

**Comment by connorff** [comment]
Currently supports testing forward and backward select. I resolved the issue with `flowistry.<command>` not being accessible when running locally so I didn't need to ping the available vscode commands.

A couple things not yet implemented/undecided on how to best move forward:

- What to test: I was going to copy over the tests from the Rust unit tests, but 
    - There are a lot of backward slice tests and copying them over by hand takes a while (could write a script to do it but that adds other complexities)
    - Because, this PR just tests the ability for a slice performed on a value to select the correct text, duplicating the Rust unit tests might be overkill
- Generating files
    - The current implementation doesn't store any values generated from the call to `cargo flowistry`, so it's called on every test run
    - Storing the slice output would be pretty simple, but I wanted to see if writing it to a file is something we'd find useful

File: ide/src/tests/unit/index.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    const testsRoot = __dirname;
+
+    return new Promise((resolve, reject) => {
+        glob('**/**.test.js', { cwd: testsRoot }, (err, files) => {
+            if (err) {
+                return reject(err);
+            }
+
+            // Add files to the test suite
+            files.forEach(f => mocha.addFile(path.resolve(testsRoot, f)));
```
[the remark this answers is not in the document]
Comment (connorff) [reply ide/src/tests/unit/index.ts]: I expected `mocha.addFile` to ignore the index and array arguments passed from the `forEach` callback, but it seems like it doesn't. For example, this works:
```js
files.map(f => path.resolve(testsRoot, f)).forEach(f => mocha.addFile(f));
```
But this doesn't:
```js
files.map(f => path.resolve(testsRoot, f)).forEach(mocha.addFile);
```

File: ide/src/tests/slice/select.test.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    });
+  });
+
+  test("backward select", async () => {
+    const backward_selections = await resolve_sequentially(backward_slices, get_slice_selections);
+
+    backward_selections.forEach(async (selection) => {
+      expect(selection.expected_selections).to.be.deep.equalInAnyOrder(selection.actual_selections);
+    });
+  });
```
Comment (connorff) [inline ide/src/tests/slice/select.test.ts:27]: I split the [previous method](https://github.com/willcrichton/flowistry/blob/5c805f25142b8d27bfd069601709c477657ba5d1/ide/src/tests/unit/extension.test.ts#L94-L103) for testing the slice selections into two separate `test`s because this seems more readable and removes the code-smell of having to create a useless `it()`.

Let me know if you preferred the previous method though, since it did reduce code duplication.

File: ide/src/tests/install/install.test.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+import { expect } from "chai";
+import vscode from "vscode";
+import waitForExpect from "wait-for-expect";
+
+const flowistryCommandsExist = async () => {
+    const commands = await vscode.commands.getCommands();
+    return commands.filter((command) => command.includes('flowistry.')).length > 0;
+};
+
+suite("Flowistry installation tests", () => {
```
Comment (connorff) [inline ide/src/tests/install/install.test.ts:11]: Is there something you'd like to test in addition to the Flowistry commands being available? I was thinking either:

- Expect `cargo flowistry -V` returns a valid version and doesn't fail
- Run a Flowistry command in VSCode and expect it doesn't fail

Comment (connorff) [reply ide/src/tests/install/install.test.ts:11]: Would it also be valuable to have a test that blocks requests to the release binaries causing the extension to compile from source?

File: .github/actions/create-build/action.yml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+      with:
+        toolchain: nightly-2021-09-23
+        components: rust-src, rustc-dev, llvm-tools-preview
+        target: ${{ inputs.target }}
+        profile: minimal
+        override: true
+    - uses: actions-rs/cargo@v1
+      with:
+        command: build
+        args: --release
```
Comment (connorff) [inline .github/actions/create-build/action.yml:19]: I created this [`composite action`](https://docs.github.com/en/actions/creating-actions/creating-a-composite-action) since the `build` job in the `tests` action use the same build steps as the `release` action. If it looks good, it might be good to reuse this in `release.yml` so both actions build the same.

File: .github/workflows/tests.yml [hunk 5]
```diff
@@ -9,11 +9,96 @@ on:
...
+    runs-on: ${{ matrix.os }}
+
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu-latest]
+
+    steps:
+      - uses: actions/checkout@v2
+      - run: cargo test
```
Comment (willcrichton) [inline .github/workflows/tests.yml:41]: It's good to use the cargo action instead of the `cargo test` shell command since (IIRC) GitHub can show the output in a more structured way.

File: .github/workflows/tests.yml [hunk 6]
```diff
@@ -9,11 +9,96 @@ on:
...
+        working-directory: ide
+      - run: xvfb-run -a npm run test -- install
+        working-directory: ide
+
+  install_head:
+    name: Install from HEAD
+    needs: build
+    runs-on: ${{ matrix.os }}
+    env:
+      NODE_TLS_REJECT_UNAUTHORIZED: 0
```
Comment (willcrichton) [inline .github/workflows/tests.yml:86]: Can you leave a comment for what this is and why it's needed?

File: .github/workflows/tests.yml [hunk 7]
```diff
@@ -9,11 +9,96 @@ on:
...
+        include:
+          - os: ubuntu-latest
+            target: x86_64-unknown-linux-gnu
+
+    steps:
+      - uses: actions/checkout@v2
+      - uses: ./.github/actions/create-build
+        with:
+          target: ${{ matrix.target }}
+      - uses: actions/upload-artifact@v2
```
Comment (willcrichton) [inline .github/workflows/tests.yml:26]: This is only needed to share artifacts with the `install_head` job right? 

If so, it might just make sense to fuse these two jobs together.

File: ide/src/tests/install/install.test.ts [hunk 8]
```diff
@@ -0,0 +1,22 @@
+import { suite, test } from "mocha";
+import { expect } from "chai";
+import vscode from "vscode";
+import waitForExpect from "wait-for-expect";
+
+const flowistryCommandsExist = async () => {
+    const commands = await vscode.commands.getCommands();
+    return commands.filter((command) => command.includes('flowistry.')).length > 0;
```
Comment (willcrichton) [inline ide/src/tests/install/install.test.ts:8]: `command.startsWith`

Comment (willcrichton) [reply ide/src/tests/install/install.test.ts:8]: `return commands.find(command => command.startsWith('flowistry.'));`

File: ide/src/tests/slice/select.test.ts (hunk 2 above)
Comment (willcrichton) [reply ide/src/tests/slice/select.test.ts:27]: Yeah since the code is identical I would still factor out a function. You could use a curried function to make it ✨ beautiful ✨ 

```js
const slice_test = slices => async () => {
  const selections = await resolve_sequentially(slices, get_slice_selections);
  selections.forEach(...)
}

test("forward select", slice_test(forward_slices));
```

File: ide/src/tests/slice/util/slice_helpers.ts [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+import { to_vsc_range } from "../../../vsc_utils";
+import { TestSlice } from "../mock_data/slices";
+import { MOCK_PROJECT_DIRECTORY } from "../../constants";
+
+export declare const TOOLCHAIN: {
+    channel: string;
+    components: string[];
+};
+
+const LIBRARY_PATHS: Partial<Record<NodeJS.Platform, string>> = {
```
Comment (willcrichton) [inline ide/src/tests/slice/util/slice_helpers.ts:14]: Can we share the common logic here with `setup.ts`?

File: ide/src/tests/slice/util/slice_helpers.ts [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+/**
+ * Merge overlapping VSCode Ranges.
+ * @param ranges Array of ranges to merge.
+ * @returns Array of merged ranges.
+ */
+const merge_ranges = (ranges: vscode.Range[]): vscode.Range[] => {
+    const merged_ranges = [ranges[0]];
+
+    ranges.slice(1).forEach((range) => {
+        const last_range = merged_ranges[merged_ranges.length - 1];
```
Comment (willcrichton) [inline ide/src/tests/slice/util/slice_helpers.ts:92]: You can use `_.last(merged_ranges)!`

File: ide/build.cjs [hunk 11]
```diff
@@ -25,7 +25,7 @@ let common = {
 };
 
 let extension = estrella.build({
-  entryPoints: ["src/extension.ts"],
+  entryPoints: ["src/extension.ts", "src/tests/slice/util/slice_helpers.ts"],
```
Comment (willcrichton) [inline ide/build.cjs:28]: I don't think we want to have `slice_helpers.ts` be an entryPoint in the same build. This would include `slice_helpers.js` in the distribution, but it's only used for development.

I would either conditionally add it as an entry point, or add another `estrella.build` call with a different `outdir` that isn't `dist`.

File: ide/src/tests/runTests.ts [hunk 12]
```diff
@@ -0,0 +1,48 @@
+import path from 'path';
```
Comment (willcrichton) [inline ide/src/tests/runTests.ts:1]: Since the tests are split across so many files, it would be good to have a document summarizing the test architecture. This could either be a comment at the top of this file, or a standalone `src/tests/README.md`. Just a paragraph or two explaining the directory structure and how data flows between each component.

**Review by willcrichton — CHANGES_REQUESTED** [review]
Almost there! Thanks for all your hard work.

File: ide/src/tests/install/install.test.ts (hunk 3 above)
Comment (willcrichton) [reply ide/src/tests/install/install.test.ts:11]: My first instinct was that both of those things should already be tested by running the slicer. But I see that those tests don't get executed in the `install_head` / `install_release` jobs.

Is the reason we don't just do `npm test` in those jobs for efficiency?

Comment (willcrichton) [reply ide/src/tests/install/install.test.ts:11]: > Would it also be valuable to have a test that blocks requests to the release binaries causing the extension to compile from source?

It's enough of an edge case that we don't need to worry about it for now.

Comment (connorff) [reply ide/src/tests/install/install.test.ts:11]: > Is the reason we don't just do npm test in those jobs for efficiency?

I initially separating them to:

- Make it quicker to run tests by
    - skipping testing the latest release since it should already have been tested and
    - running the `install` and `slice` tests in parallel
- Make it clear whether the extension failed because it installed incorrectly or because the extension/driver is broken

That being said, it might be beneficial to test after install since I could remove `test-vscode` and run the tests after install. Another benefit would be removing the `install-build` composite action thingy since `install_head` and `install_release` will do that for us.

Whatever seems best works for me!

File: .github/workflows/tests.yml (hunk 7 above)
Comment (connorff) [reply .github/workflows/tests.yml:26]: It's also used for the `vscode-test` job but if we remove that it would make sense to fuse them.

Comment (connorff) [reply .github/workflows/tests.yml:26]: The artiface is used in the `test-vscode` job but if we remove that we can fuse these two together.

File: ide/src/tests/unit/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (willcrichton) [reply ide/src/tests/unit/index.ts]: Love to program functionally in Javascript, a functional programming language

File: ide/src/tests/install/install.test.ts (hunk 3 above)
Comment (willcrichton) [reply ide/src/tests/install/install.test.ts:11]: For now, I think let's just keep it simple. Let's do one job in the following sequence:

1. `cargo build`
2. copy binaries into a distribution zip
3. `cargo test`
4. `npm build`
5. `npm test --install <the zip>`
 
This loses the parallelism between cargo and vscode tests, but that shouldn't be a big issue because the cargo tests (at least) are very fast. Certainly faster than spinning up a new VM. I don't think testing vscode with pre-installed binaries is important -- just testing the "install from server" path is fine.
